src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -53,135 +53,159 @@ namespace impl {
 
   // We need to use a temporary, since we don't know whether result refers to a thread-local
   // variable (e.g., automatic variable) or to shared-memory (e.g., an entry of a view).
   // Hence, perform calculations on a local var, then copy back into the output result.
-  ValueType temp = result;
-
-  // Note: this team barrier is needed in some extreme case. Without it, it *could* happen that,
-  //       if result is a ref to shared mem (e.g., an entry of a view) rather than thread-local,
-  //       one team member might reach the end of the fcn (hence, updating result) *before*
-  //       another thread might have the chance to init temp.
-  team.team_barrier();
+  ValueType result = Kokkos::reduction_identity<ValueType>::sum();
 
   // Perform a packed reduction over scalar indices
-  const bool has_garbage_begin = begin%vector_size != 0;
-  const bool has_garbage_end   = end%vector_size != 0;
-  const int  pack_loop_begin   = (has_garbage_begin ? begin/vector_size + 1 : begin/vector_size);
-  const int  pack_loop_end     = end/vector_size;
+  const bool has_garbage_begin = begin % N != 0;
+  const bool has_garbage_end   =   end % N != 0;
+  const int  pack_loop_begin   = (has_garbage_begin ? begin/N + 1 : begin/N);
+  const int  pack_loop_end     = end/N;
 
   // If first pack has garbage, we will not include it in the below reduction,
   // so manually add the first pack (only the non-garbage part)
   if (has_garbage_begin) {
-    const PackType temp_input = input(pack_loop_begin-1);
-    const int first_indx = begin%vector_size;
-    Kokkos::single(Kokkos::PerThread(team),[&] {
-      for (int j=first_indx; j<vector_size; ++j) {
-        temp += temp_input[j];
+    const auto temp_input = input(pack_loop_begin-1);
+    const int first_indx = begin % N;
+    Kokkos::single(Kokkos::PerThread(team),[&] (ValueType& result) {
+      for (int j=first_indx; j<N; ++j) {
+        result += temp_input[j];
       }
-    });
+    },result);
   }
 
-  // Complete packs to be reduced. If Serialize, reduce at pack the level, then
-  // sum into result. Else, sum up packs, then sum over resulting pack into result.
+  // Complete packs to be reduced. If Serialize, reduce each pack first, then
+  // sum into result. Else, sum up packs, then reduce the packed sum into result.
   if (pack_loop_begin != pack_loop_end) {
     if (Serialize) {
-      impl::parallel_reduce<Serialize>(team, pack_loop_begin, pack_loop_end,
-                                       [&](const int k, ValueType& local_sum) {
-          // Sum over pack entries and add to local_sum
-          ekat::reduce_sum<Serialize>(input(k),local_sum);
-      }, temp);
+      impl::parallel_reduce<Serialize,ValueType>(
+          team, pack_loop_begin, pack_loop_end,
+          [&](const int k, ValueType& local_sum) {
+            // Sum over pack entries and add to local_sum
+            ekat::reduce_sum<Serialize>(input(k),local_sum);
+      },result);
```

Comment:
Here I believe you're relying on `impl::parallel_reduce` to _not_ 0-init `result`. I think this might be working b/c of a sneaky alignment in assumptions: The `Serialize` code path _doesn't_ 0-init `result` and you rely on that here, while the `not Serialize` code path _does_ 0-init `result`, but the code below is OK with that.

Reply:
Yeah, I was relying on that. I will mod `impl::parallel_reduce` so that the input result is not lost, regardless of Serialize